### PR TITLE
feat: improve error responses

### PR DESCRIPTION
**Pull request by ricellis**
## PR summary

Add event hook for error response augmentation

Fixes: s1004

**Note: An existing issue is [required](https://github.com/IBM/cloudant-python-sdk/blob/main/CONTRIBUTING.md#PRs) before opening a PR.**

## PR Checklist

Please make sure that your PR fulfills the following requirements:

- [x] The commit message follows the
[Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type  

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [ ] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## What is the current behavior?

Cloudant/CouchDB error models are handled by the core with only the `error` field being processed. Obtaining e.g. the `reason` field requires additional user code.

## What is the new behavior?

The error response is augmented with `errors` and `trace` properties.
This improves the way the core handles the error response and enhances the user facing exception string to include the `reason`.
The `trace` is the CouchDB request ID which can be used to correlate with logs.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

The existing error model fields are retained, the new fields are backwards compatible.
Exception strings will be different, but the types remain the same.

## Other information

Create _error_response_hook function to augment
error messages.
Add hook to the client requests session.
Unit tests for the hook.

### CODE REVIEW

**Review by emlaver — COMMENTED** [review]
Looks good! A couple comments

File: test/unit/test_cloudant_base_error_augment.py [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+                    method,
+                    url=self._base_url + f'/{self._db_name}/{self._doc_id}',
+                    content_type=mock_content_type,
+                    **mock_response
+                )
+                if expect_raises:
+                    with self.assertRaisesRegex(ApiException, expected_response['message']) as e_ctx:
+                          self._make_request(method, mock_response['stream'])
+                    caught_exception = e_ctx.exception
+                    actual_status_code = caught_exception.status_code
```
[the remark this answers is not in the document]
Comment (ricellis) [reply test/unit/test_cloudant_base_error_augment.py:127]: Other way round I think https://github.com/IBM/python-sdk-core/blob/ea2d73345dbb882a051308b84def60738df6a225/ibm_cloud_sdk_core/api_exception.py#L57

File: test/unit/test_cloudant_base_error_augment.py [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+                mock_content_type = mock_response['headers'].pop('content-type', None)
+                mock_ctx.add(
+                    method,
+                    url=self._base_url + f'/{self._db_name}/{self._doc_id}',
+                    content_type=mock_content_type,
+                    **mock_response
+                )
+                if expect_raises:
+                    with self.assertRaisesRegex(ApiException, expected_response['message']) as e_ctx:
+                          self._make_request(method, mock_response['stream'])
```
[the remark this answers is not in the document]
Comment (ricellis) [reply test/unit/test_cloudant_base_error_augment.py]: Done and squashed into 37c61ab

File: test/unit/test_cloudant_base_error_augment.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (emlaver) [reply test/unit/test_cloudant_base_error_augment.py:127]: Yes you're right, I was reviewing locally with an old SDK core version.

File: test/unit/test_cloudant_base_error_augment.py [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+        self.assertEqual(_error_response_hook, self._service.get_http_client().hooks['response'][0])
+
+    def test_response_hook_not_double_added_via_set_http_client(self):
+        client = self._service.get_http_client()
+        self._service.set_http_client(client)
+        response_hooks =  self._service.get_http_client().hooks['response']
+        self.assertEqual(1, len(response_hooks))
+        self.assertEqual(_error_response_hook, response_hooks[0])
+
+    def test_response_hook_added_via_set_http_client(self):
```
Comment (mojito317) [inline test/unit/test_cloudant_base_error_augment.py:179]: This test seem to me a duplication of `test_response_hook_not_double_added_via_set_http_client`.

Comment (mojito317) [reply test/unit/test_cloudant_base_error_augment.py:179]: right, I missed that line!

**Review by emlaver — APPROVED** [review]

**Review by mojito317 — APPROVED** [review]

**Pull request merged**